### ISSUE TITLE
[Deferred] SDL 0019 - iOS Library Proxy Layer Should Not Be Public

### ISSUE THREAD

**Issue by theresalech** [smartdevicelink/sdl_evolution#48]
Hello SDL community,

The review of the revised proposal "iOS Library Proxy Layer Should Not Be Public" begins now and runs through February 7, 2017.  The original review of "iOS Library Proxy Layer Should Not Be Public" occurred October 5, 2016 through October 19, 2016. The proposal is available here:

https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0019-ios-proxy-layer-nonpublic.md

Reviews are an important part of the SDL evolution process. All reviews should be sent to the associated Github issue at:

https://github.com/smartdevicelink/sdl_evolution/issues/48

What goes into a review?

The goal of the review process is to improve the proposal under review through constructive criticism and, eventually, determine the direction of SDL. When writing your review, here are some questions you might want to answer in your review:
- Is the problem being addressed significant enough to warrant a change to SDL?
- Does this proposal fit well with the feel and direction of SDL?
- If you have used competitors with a similar feature, how do you feel that this proposal compares to those?
- How much effort did you put into your review? A glance, a quick reading, or an in-depth study?
  Please state explicitly whether you believe that the proposal should be accepted into SDL.

More information about the SDL evolution process is available at

https://github.com/smartdevicelink/sdl_evolution/blob/master/process.md

Thank you,
Theresa Lech

Program Manager - Livio
theresa@livio.io

**Comment by theresalech** [comment]
This proposal has been deferred to next week.  We will need more information about the effort required by developers to upgrade their app to use the iOS 4.3.0 APIs, perhaps an upgrade guide.

**Comment by theresalech** [comment]
This proposal has been deferred for next week. We need a comparison of how public methods in the proxy layer will be replaced with the new APIs. PutFile streaming needs a proposal on how the FileManager will handle it.

**Comment by joeljfischer** [comment]
I would like to note that with the new updates to this proposal, PutFile streaming isn't explicitly addressed, but there is a new proposal on how to handle it that is awaiting review. 

https://github.com/smartdevicelink/sdl_evolution/pull/74

**Comment by tenajrm** [comment]
Is it backward compatibility I can still using SDLProxy? or does the new upgrade to SDL 5.0  will force developers to implement SDLManager and then replace all my SDLProxy implementation? and  is there any SDL_core  affectation for this change?

**Comment by joeljfischer** [comment]
In the SDL 4.x branches you can still use `SDLProxy`. This proposal is such that in v5.0.0, `SDLProxy` would be removed for all reasons stated in the proposal. So yes, this would force any developers not yet using `SDLManager` to do so. This changeover is not especially difficult at a minimal level, as described in the proposal. This would not affect SDL Core in any way.

**Comment by kshala-ford** [comment]
The proposal shouldn't contain code copied from the mobile weather example. Either all MobileWeather specific code is removed or the proposal shows a proper replacement. You're actually comparing
```objc
- (void)setupProxy {
   if ([self proxy] == nil) {
       [self setProxy:[SDLProxyFactory buildSDLProxyWithListener:self]];
   }
}

- (void)teardownProxy {
   if ([self proxy] != nil) {
       [[self proxy] dispose];
       [self setProxy:nil];
   }
}

- (void)onProxyClosed {
   [self teardownProxy];
   [self setupProxy];
}

- (void)onProxyOpened {
   [self registerApplicationInterface];
}

- (void)registerApplicationInterface {
   SDLRegisterAppInterface *request = [[SDLRegisterAppInterface alloc] initWithAppName:@"MobileWeather" appId:@"12345" languageDesired:[SDLLanguage EN_US]];
   [self sendRequest:request];
}

- (void)onRegisterAppInterfaceResponse:(SDLRegisterAppInterfaceResponse *)response {
}
```

with following code

```objc
SDLLifecycleConfiguration *lifecycleConfig = [SDLLifecycleConfiguration defaultConfigurationWithAppName:@"MobileWeather" appId:@"330533107"];
lifecycleConfig.logFlags = SDLLogOutputConsole;

SDLConfiguration *config = [SDLConfiguration configurationWithLifecycle:lifecycleConfig lockScreen:[SDLLockScreenConfiguration enabledConfiguration]];

self.manager = [[SDLManager alloc] initWithConfiguration:config delegate:self];

[self.manager startWithReadyHandler:^(BOOL success, NSError * _Nullable error) {
}
```

The method to send RPCs is called `sendRequest` and provides a completion handler that should be mentioned.

Besides the content of the proposal file I have some more questions to the managers:
- In how far interfers the lifecycle manager with a ChangeRegistration request?
- How do I know if language mismatch occurred?
- How do I know if the "resume hash" was accepted?
- Why is the WiFi transport layer called "debug"? Couldn't WiFi be used in production?
- Whats the readyHandler? What does "ready" mean? And why isn't it placed inside SDLManagerDelegate?
- Where's the inline documentation of all new classes? (Most developers may not understand the purpose of the classes) 
- What's an artwork and when do I use a SDLImage or a UIImage?

I do understand that the manager(s) makes things easier but before making the proxy private and not accessible for an app developer there should be a stable and mature replacement which is well thought.

**Comment by joeljfischer** [comment]
@kshala-ford 

> The proposal shouldn't contain code copied from the mobile weather example. Either all MobileWeather specific code is removed or the proposal shows a proper replacement.

I'm not sure what is meant by this. MobileWeather is an open-source SDL project, and including code comparing implementations seems legitimate to me. I'm not sure why you don't think comparing those two pieces of code is illegitimate either. Between the two implementations they are equivalent.

If you do not like my use of MobileWeather, Alex also updated HelloSDL. You can see the majority of the [old SDL code](https://github.com/smartdevicelink/hello_sdl_ios/blob/7ec7703edf3aeb0c330971de26b7d14440321762/HelloSDL/HelloSDL/HSDLProxyManager.swift) compared to the [new SDL code](https://github.com/smartdevicelink/hello_sdl_ios/blob/feature/swift-3.0/HelloSDL/HelloSDL/HSDLProxyManager.swift). There are Objective-C implementations as well: [old](https://github.com/smartdevicelink/hello_sdl_ios/blob/c3fb9e7bf6e93898abdd34b162e656886e421443/HelloSDL/HelloSDL/HSDLProxyManager.m) and [new](https://github.com/smartdevicelink/hello_sdl_ios/blob/develop/HelloSDL/HelloSDL/HSDLProxyManager.m). Idiomatic use of the manager frameworks more than halved the amount of code required to perform the same functionality.

This proposal however, has been focused on the *minimum* changes necessary, and you will see that reflected in the proposal.

> The method to send RPCs is called sendRequest and provides a completion handler that should be mentioned.

You are correct that the manager sending method is called `sendRequest:` not `sendRPC:` as is mentioned in the proposal, but there are two methods, one with a completion handler, and one without. Only the naming was incorrect here.

> Besides the content of the proposal file I have some more questions to the managers:

Besides the content of the proposal, you have a number of questions. While I would be happy to answer them, they are all entirely outside of the scope of this review and proposal (especially since the majority apply to old and new code equally), and I won't respond to them here. Performing any function is no more difficult or less possible (though it may be different) with `SDLManager` than it was with `SDLProxy`. If you would like these questions answered, I would be happy to answer them over Slack or email.

> I do understand that the manager(s) makes things easier but before making the proxy private and not accessible for an app developer there should be a stable and mature replacement which is well thought.

Perhaps you don't like `SDLManager` or believe it is well thought out, but it was the product of nearly a year of architecture and code work, and it is the replacement to `SDLProxy`. It is mature, and is in use by a number of developers. Keeping `SDLProxy` indefinitely will harm the project's feature and bug fix development and severely limit our ability to update the internals of SDL. As it stands, any change to the backing layers for `SDLManager` are breaking changes due to unfortunate architecture decisions made long ago. I truly believe this change is necessary for the health of the project. Even discarding much of the "new" of the `SDLManager` API, everything that was possible with `SDLProxy` remains possible, as shown in the proposal.

**Comment by kshala-ford** [comment]
No offense @joeljfischer. this is no critism to all the work on the managers. I'm trying to be objective.

All I wanted to say about the proposal is that it contains mistakes and that it does not provide a realistic code comparison. A propsal of such consequences should be written very carefully.

I put that list of questions knowing that it is out of scope because I wanted to point out a very important fact that the proposal is not mentioning: As far as I can see in the latest vesrion the manager lacks support of hash resumption and localization (language (mis)match). Especially the second point is fundamental for european apps with all the different languages. 

Yes you're right. My concerns are out of scope of this proposal. But I don't see any benefits for an app developer by hiding the proxy. I just see more work for the app developers forced by the SDK and I see demands to add missing features and fix issues that didn't exist in the past. 

In the past in open source projects I've seen many APIs being the successor for older techniques. The new APIs were always competing with the existing ones. Constant improvements increased the acceptance of the new APIs and replaced the existing at some point. Bottom line: The users decided when to move to the successor. I would rather like to see users to accept the new managers by themself and to decide when to use them.

I would say yes to proposals that simplify the developing against and integrating SDL. Yes the managers simplify many things that were painful in the past. At the end of the day the app developer needs less lines of code to code a good user experience which is fantastic. I just don't think the manager is at this point to replace the proxy.

I don't want to hinder the process. My recommendation is to say **yes under the condition** of improvements.

**Comment by joeljfischer** [comment]
There has been a typo fix in this proposal; see [this commit](https://github.com/smartdevicelink/sdl_evolution/commit/78753722298cb34969ab1ba4625c4ca9c0380281) for more details.

**Comment by joeljfischer** [comment]
@kshala-ford I'm not taking offense here, I am simply trying to defend my proposal, because I do believe it's the right course of action.

I have fixed the minor typo. I responded with why it is a realistic code comparison, and I'm still not sure why you think it's not. Those two sections of code do the same thing.

The manager classes do provide hash resumption, you can watch for the hash notification, and on next start, set `SDLLifecycleConfiguration`'s `resumeHash` property. As far as I can tell, although the manager does not currently automatically help with language, it is possible to unregister and reregister with a new language if desired. Both of these could be automated, but they are both still manually possible.

You are correct in that it does not benefit the app developer to hide the proxy. Deprecating or removing APIs will never help developers who use those APIs. However, due to the decisions that were made when the initial build of this project was made around 6 years ago, we cannot improve the internals of the library without making major version changes. This has restricted us enormously by preventing those internals, for example, from being unit testable. If we do not make this change now, we will continue to be restricted until the next major version change, and I guarantee there will still be developers who have not moved on.

If necessary, we can back-port major bug fixes into the 4.X line for a short time if the committee desires to continue to pay for such work, while simultaneously working primarily on the 5.X line.

As I stated before, I do not believe there is anything impossible or significantly more difficult for a developer to do now compared with using `SDLProxy`. If you find anything, please let us know by filing an issue.

Finally, a yes under condition of improvements is really a no, or a deferral. It would also be helpful for you to state what improvements you think are necessary.

**Comment by mrapitis** [comment]
I understand the need to provide better and more developer friendly API's, however my concern with this proposal is that we are removing the proxy library interface before giving partners and developers much of a chance to upgrade their apps and provide feedback.  To my knowledge there are only a couple of sample apps currently utilizing the new API's -- we may also need to beef up documentation?  I think it may be too early to remove the deprecated proxy from the ios library.

**Comment by joeljfischer** [comment]
Hi @mrapitis, all sample apps i'm aware of are using the new apis, and there will be a load of documentation guides releasing very soon. Regarding timing, by the time 5.0 released with this change, it will have been at least 6 months since v4.3 released officially. I understand the concern, but these classes are currently totally untestable and will continue to be until we can totally refactor them. The best way to do this will be to hide this layer. A minimal update for developers as laid out in this proposal should not be difficult. Also, as I mentioned above, critical bug fixes could continue to be applied to the v4.X line for a short time to enable developers a longer runway to update.

**Comment by kshala-ford** [comment]
@joeljfischer could you please explain what classes are not testable and what's the issue behind testing it? 

The managers are basically a layer (application layer) on top of the RPC layer. Hiding all layers except the application layer doesn't make it obsolete testing them. Issues in the RPC layer will cause issues in the application layer.

I'm sorry but I have to disagree to "A minimal update for developers as laid out in this proposal should not be difficult". 
- The proxy must be replaced.
- Every implemented method of the listener protocol has to be moved
- Every notification method of the listener has to be manually registered 
   - with no support of auto complete. 
   - with the need to cast to the right type

Migration is managable for sure but takes days for production apps.

**Comment by asm09fsu** [comment]
@kshala-ford I have assisted a few groups of individuals in updating their proxies, as well as [starring a post](https://smartdevicelink.slack.com/archives/ios-help/p1483461388000341) in Slack relating to updating from pre-4.3 (using `SDLProxy`) and post-4.3 (using `SDLManager`), and [how to listen to the new notifications](https://smartdevicelink.slack.com/archives/ios-help/p1479748942000204).

Yes, this new style requires a bit of learning, but so does every API. The iOS documentation has a [pending pull request](https://github.com/smartdevicelink/sdl_ios_guides/pull/3) that Joel and I are working on where I have rewritten all of the documentation to use these new classes and explain how to use them.

From my discussions with developers, this does not take them long to update. From my interactions it takes anywhere from 1-3 days, and 3 days is really pushing it. Only allowing developers to connect via `SDLManager` means we know that every app connecting will connect the same way, and make it easier for us to fix bugs and mitigate integration issues.

Removing the proxy is going to be necessary moving forward, as Joel has stated, in allowing us to more easily modify the internal workings of the proxy, as these haven't really been updated in ~2 years.

The replacement of the delegate callbacks and notifications really is not tedious. It would take a developer at the most an hour or so to update.

**Comment by joeljfischer** [comment]
@kshala-ford 
> Hiding all layers except the application layer doesn't make it obsolete testing them. Issues in the RPC layer will cause issues in the application layer.

This is what I am saying. What you refer to as the RPC layer (in reality, 3 layers, the transport, protocol, and proxy layers), is currently untestable (I'm talking in terms of unit tests). These layers are all tied together publicly in a way that requires major version changes to break. They all were poorly designed, certainly without testing in mind, around 5 years ago, and have had bolt-on additions over that span. They simply cannot be tested in any meaningful way without an overhaul of their API and architecture. Doing so either requires a major version change and rewrite, or to make them private in which case they can be restructured as needed without further version changes (outside of hotfixes). These layers (outside of the proxy layer, which publicly exposes and ties them together) were never intended for developer use, but we cannot remove them without altering the proxy layer.

As Alex says, an update is really quite easy. The MobileWeather app contained ~1800 lines of SDL code and I updated it in ~2-3 hours.

**Comment by theresalech** [comment]
While this is a goal for the iOS library, it's been determined that more time is needed to make the library private.  This proposal has been deferred until teams have a communication plan in place to notify and assist app partners with this change.

**Comment by theresalech** [comment]
This proposal has been accepted by the Steering Committee.  The committee has agreed that this should be implemented after 5.0 and will discuss further when planning the 6.0 roadmap/release.

**Comment by theresalech** [comment]
Issue Entered: [[SDL 0019] iOS Library Proxy Layer Should Not Be Public](https://github.com/smartdevicelink/sdl_ios/issues/603)
